### ISSUE TITLE
Serialise fails to round-trip all characters

### ISSUE THREAD

**Issue by bgamari** [well-typed/cborg#135]
Surrogates fail to round-trip.

```
λ⟩ (deserialise . serialise) '\55296' :: Char
'\65533'
λ⟩ all (('\xFFFD' ==) . deserialise . serialise) ['\xD800'..'\xDFFF']
True
```

**Comment by bgamari** [comment]
One design here would be able add a `Utf8ByteArray :: ByteArray -> Encoding` constructor to `Encoding` (and the same for `DecoderAction`). The same should be done for bytes anyways, per #136.

@dcoutts, thoughts?

**Comment by bgamari** [comment]
Adding `ByteArray` into the mix is a bit unfortunate as it lacks necessary instances (`Show`, `Eq`, `Ord`). Arguably these should be introduced upstream, but we would likely need to provide them ourselves if presented with an old `primitive` release. Alternatively we could introduce our own `ByteArray` type, although I'm not terribly fond of this option.

**Comment by hvr** [comment]
@bgamari not sure if this if any help, but would the `ShortText` type from https://hackage.haskell.org/package/text-short be any help here?

**Comment by bgamari** [comment]
As far as I can see, the viable options here are adding encoders/decoders for either,

 * `Data.Text.Short.ShortText`
 * `Data.ByteString.ShortByteString.ShortByteString`
 * `Data.Primitive.ByteArray.ByteArray` (with necessary orphan instances for `Eq`, `Ord`, `Show`)
 * a hand-rolled `ByteArray` wrapper

**Comment by dcoutts** [comment]
I'm unclear how this helps with `String` which can represent more than can be represented in UTF8.

**Comment by bgamari** [comment]
@dcoutts my thought was that users who anticipate leaving the representable range of `Text` should instead give us UTF-8. Additionally we can also change the `Char` instance to not go through `Text` but instead directly encode to UTF-8.

**Comment by bgamari** [comment]
After doing a bit of reading, it's actually not entirely clear to me how to deal with this. While `Char` can be easily represented as a word, the case a `String` (which may contain invalid surrogates) is quite thorny. UTF-8 explicitly states that encodings corresponding to surrogate halves  should be treated as an invalid byte sequence. The more liberal [WTF-8](https://simonsapin.github.io/wtf-8/#intended-audience) explicitly states that the format should not be used for interchange, rendering it a poor fit for our application.

**Comment by bgamari** [comment]
Alright, I thought about this a fair bit.

Ultimately, this whole issue comes down to the fact that `Char` is defined to be a Unicode code-point, but admits values which aren't representable in UTF-8. However, I agree that despite this quirk we should strive to ensure that round-tripping works properly. Consequently, I think the most reasonable way forward is to change the `String` instance to do two-passes: first we scan over the string looking for unrepresentable code-points. If we find such characters then we encode the thing in a bytes value as either UTF-32 or generalized UTF-8 (as defined by the WTF-8 specification). In the event that the string is clean, then we retain the current UTF-8 text encoding.

Does this sound reasonable?

**Comment by hvr** [comment]
> If we find such characters then we encode the thing in a bytes value as either UTF-32 

Nope! :-) You can't call it UTF-32 in that case, since, because ISO/IEC 10646:2014 Clause 9.4 says

> Because surrogate code points are not UCS scalar values, UTF-32 code units in the range 0000 D800-0000 DFFF are ill-formed 

Obviously, code units above 0x10FFFF are ill-formed as well.

Also, well-formed UTF-16 wouldn't allow arbitrary 16-bit code units either.


It could be argued, that according to the Haskell 2010 specification, which states things like

> The character type Char is an enumeration whose values represent Unicode **characters**

> The character type Char is an enumeration whose values represent Unicode (or equivalently ISO/IEC
10646) **characters** (see http://www.unicode.org/ for details).

would legitimate GHC to disallow codepoints in the U+D800 to U+DFFF range, as they don't "exist" at the character level of Unicode; they're merely an implementation detail of UTF16, and are almost exclusively forbidden to be used in any other context according to the Unicode specification (which Haskell 2010 is based upon).

**Comment by dcoutts** [comment]
Ok, how about this: we do a one-pass conversion into a UTF-8/WTF-8 byte string (in some rep) where as part of the result we keep track of whether there were any out of range code points. If there were none then we pass the valid UTF-8 byte string off to the writer layer to include as a CBOR text. If there were invalid code points then we pass the byte string off to the writer layer to include as a CBOR bytestring. We may wish to declare and use a CBOR tag for WTF-8.

This will indeed require that we add a new encoder primitive for CBOR text that does not use the `Text` type but a byte string (`ByteArray#` ?) that is presumed to be valid UTF-8. This would then be used by instances for String, Text etc.

For generality we may require both possibly-unpinned style `ByteArray#` and also `Ptr` based encoder primitives. There may still be value in special cases for things like `Text` by allowing the conversions to be moved to the interpreter rather than duplicated at all call sites.

**Comment by verement** [comment]
I like @dcoutts’s approach, and it’s one I’ve been suggesting to @bgamari in email. I would just point out that “WTF-8” is not a good name for the encoding, because it specifically excludes surrogate code point pairs, which make it ill-formed. The correct name for the encoding I believe would be “generalized UTF-8” which is a superset of both UTF-8 and WTF-8.

**Comment by hvr** [comment]
@verement Indeed, I should have called it "generalized UTF-8" when I spoke to @bgamari; to my defense, I was misled by Wikipedia's description of [WTF-8](https://en.wikipedia.org/wiki/UTF-8#WTF-8) which made it sound to me as if WTF-8 would simply lift the restriction... only after reading the actual WTF-8 spec I realised that it's a more complicated encoding...

**Comment by bgamari** [comment]
If I'm not mistaken this is what I implemented in #139.
